src/compiler/diab.rs

```diff
@@ -0,0 +1,694 @@
+    parsed_args: &ParsedArguments,
+    cwd: &Path,
+    env_vars: &[(OsString, OsString)],
+    _may_dist: bool,
+) -> SFuture<process::Output>
+where
+    T: CommandCreatorSync,
+{
+    let mut cmd = creator.clone().new_command_sync(&executable);
+    cmd.arg("-E")
+        .arg(&parsed_args.input)
+        .args(&parsed_args.preprocessor_args)
+        .args(&parsed_args.common_args)
+        .env_clear()
+        .envs(env_vars.iter().map(|&(ref k, ref v)| (k, v)))
+        .current_dir(cwd);
+
+    if log_enabled!(Trace) {
+        trace!("preprocess: {:?}", cmd);
+    }
+    run_input_output(cmd, None)
+}
+
+pub fn generate_compile_commands(
+    _path_transformer: &mut dist::PathTransformer,
+    executable: &Path,
+    parsed_args: &ParsedArguments,
+    cwd: &Path,
+    env_vars: &[(OsString, OsString)],
+) -> Result<(CompileCommand, Option<dist::CompileCommand>, Cacheable)> {
+    trace!("compile");
+
+    let out_file = match parsed_args.outputs.get("obj") {
+        Some(obj) => obj,
+        None => return Err("Missing object file output".into()),
+    };
+
+    let mut arguments: Vec<OsString> = vec![
+        "-c".into(),
+        parsed_args.input.clone().into(),
+        "-o".into(),
+        out_file.into(),
+    ];
+    arguments.extend(parsed_args.preprocessor_args.clone());
+    arguments.extend(parsed_args.common_args.clone());
+    let command = CompileCommand {
+        executable: executable.to_owned(),
+        arguments: arguments,
+        env_vars: env_vars.to_owned(),
+        cwd: cwd.to_owned(),
+    };
+
+    Ok((command, None, Cacheable::Yes))
+}
+
+pub struct ExpandAtArgs<'a> {
+    cwd: &'a Path,
+    stack: Vec<OsString>,
+}
+
+impl<'a> ExpandAtArgs<'a> {
+    pub fn new(cwd: &'a Path, args: &[OsString]) -> Self {
+        ExpandAtArgs {
+            stack: args.iter().rev().map(|a| a.to_owned()).collect(),
+            cwd: cwd,
+        }
+    }
+}
+
+impl<'a> Iterator for ExpandAtArgs<'a> {
+    type Item = OsString;
+
+    fn next(&mut self) -> Option<OsString> {
+        loop {
+            let arg = self.stack.pop()?;
+
+            // Just return non @ arguments
+            if !arg.starts_with("@") {
+                return Some(arg.into());
+            }
+
+            let arg = match arg.split_prefix("@") {
+                Some(arg) => arg,
+                None => return Some(arg),
+            };
+
+            // Return options that produce additional output and are not cacheable
+            if arg.starts_with("E=") || arg.starts_with("O=") || arg.starts_with("@") {
+                return Some(arg);
+            }
+
+            // According to diab [1], @file means:
+            // Read command line options from either a file or an environment
+            // variable. When -@name is encountered on the command line, the
+            // driver first looks for an environment variable with the given
+            // name and substitutes its value. If an environment variable is
```

Comment:
Wow, I thought gcc/msvc's behavior was difficult to deal with, but reading commandline options from an *environment variable* is something else!

Reply:
I really doubt that anybody uses that feature. I will check how `ccache` handles that case (yes they support `diab`).

Reply:
I couldn't find anybody here using that feature. Furthermore checked `ccache` which also doesn't support it. 
I remove the corresponding code and flag the compilation as non cacheable.

Reply:
As described above this case is now treated as a `CannotCache`.